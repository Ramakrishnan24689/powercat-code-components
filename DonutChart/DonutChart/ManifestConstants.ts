export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    Legend = 'ItemLegend',
    Key = 'ItemKey',
    Value = 'ItemValue',
    Color = 'ItemColor',
}
