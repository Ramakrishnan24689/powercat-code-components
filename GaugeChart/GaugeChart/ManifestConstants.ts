import { GaugeValueFormat } from '@fluentui/react-charting';

export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const ChartValueFormats = {
    '1': GaugeValueFormat.Fraction,
    '0': GaugeValueFormat.Percentage,
};

export const enum ItemColumns {
    Legend = 'ItemLegend',
    Key = 'ItemKey',
    Count = 'ItemCount',
    Color = 'ItemColor',
    Size = 'ItemSize',
}
