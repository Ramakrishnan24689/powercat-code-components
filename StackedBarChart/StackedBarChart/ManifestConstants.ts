export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    Title = 'ItemTitle',
    Key = 'ItemKey',
    Value = 'ItemValue',
    Color = 'ItemColor',
    Callout = 'ItemCallout',
}

export const enum InputProperties {
    SelectedKey = 'SelectedKey',
}
