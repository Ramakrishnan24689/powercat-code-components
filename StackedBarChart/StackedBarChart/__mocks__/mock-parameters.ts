/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockStringProperty, MockTwoOptionsProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        Title: new MockStringProperty(),
        HideTooltip: new MockTwoOptionsProperty(),
        BarHeight: new MockWholeNumberProperty(),
        HideLegend: new MockTwoOptionsProperty(),
        BarBackgroundColor: new MockStringProperty(),
        AccessibilityLabel: new MockStringProperty(),
        Theme: new MockStringProperty(),
        CustomColors: new MockTwoOptionsProperty(),
        items: new MockDataSet([]),
        TabIndex: new MockWholeNumberProperty(),
        Tooltip: new MockStringProperty(),
    };
}
