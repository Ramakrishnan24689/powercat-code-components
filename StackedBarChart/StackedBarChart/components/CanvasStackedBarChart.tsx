import * as React from 'react';
import { StackedBarChart, IStackedBarChartProps } from '@fluentui/react-charting';
import { IPartialTheme, ThemeProvider, createTheme } from '@fluentui/react';

export interface ICanvasStackedBarChartProps extends IStackedBarChartProps {
    themeJSON?: string;
}

export const StackedBarChartControl = React.memo((props: ICanvasStackedBarChartProps) => {
    const { themeJSON, ...rest } = props;
    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    return (
        <ThemeProvider theme={theme} style={{ backgroundColor: 'transparent' }}>
            <StackedBarChart {...rest} />
        </ThemeProvider>
    );
});
StackedBarChartControl.displayName = 'StackedBarChartControl';
