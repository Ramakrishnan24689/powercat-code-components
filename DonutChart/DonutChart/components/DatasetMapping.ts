import { DataVizPalette, IChartDataPoint, getColorFromToken } from '@fluentui/react-charting';
import { ItemColumns } from '../ManifestConstants';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    const colorArray = _colors[index % _colors.length];
    return getColorFromToken(colorArray[Math.floor(Math.random() * colorArray.length)]);
}

function getKeyWithSuffix(keyIndex: Record<string, number>, key: string): string {
    if (keyIndex[key] !== undefined) {
        keyIndex[key]++;
        return `${key}_${keyIndex[key]}`;
    } else {
        keyIndex[key] = 1;
        return key;
    }
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    customColor: boolean,
): IChartDataPoint[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        return Array.from({ length: 3 }, (_, i) => getDummyAction(`Action${i + 1}`, i));
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id, index) => {
        const record = dataset.records[id];
        const key = getKeyWithSuffix(keyIndex, record.getValue(ItemColumns.Key) as string);
        return {
            id: record.getRecordId(),
            key: key,
            legend: record.getValue(ItemColumns.Legend) as string,
            color: customColor ? getColorFromToken(record.getValue(ItemColumns.Color) as string) : _randomColor(index),
            data: record.getValue(ItemColumns.Value) as number,
        } as IChartDataPoint;
    });
}

function getDummyAction(key: string, index: number): IChartDataPoint {
    return {
        id: key,
        key: key,
        legend: key,
        color: _randomColor(index),
        data: 30,
    } as IChartDataPoint;
}
