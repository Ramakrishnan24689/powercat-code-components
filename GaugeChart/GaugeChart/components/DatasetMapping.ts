import { DataVizPalette, IGaugeChartSegment, getColorFromToken } from '@fluentui/react-charting';
import { ItemColumns } from '../ManifestConstants';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    return getColorFromToken(_colors[index][Math.floor(Math.random() * _colors[index].length)]);
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    customColor: boolean,
): IGaugeChartSegment[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('First', 1), getDummyAction('Second', 2), getDummyAction('Third', 3)];
    }
    const keyIndex: Record<string, number> = {};
    if (customColor) {
        return dataset.sortedRecordIds.map((id, index) => {
            const record = dataset.records[id];

            // Prevent duplicate keys by appending the duplicate index
            let key = record.getValue(ItemColumns.Key) as string;
            if (keyIndex[key] !== undefined) {
                keyIndex[key]++;
                key += `_${keyIndex[key]}`;
            } else keyIndex[key] = 1;
            return {
                id: record.getRecordId(),
                key: key,
                legend: record.getValue(ItemColumns.Legend) as string,
                size: record.getValue(ItemColumns.Size) as number,
                color: _randomColor(index),
            } as IGaugeChartSegment;
        });
    }
    return dataset.sortedRecordIds.map((id, index) => {
        const record = dataset.records[id];

        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            legend: record.getValue(ItemColumns.Legend) as string,
            size: record.getValue(ItemColumns.Size) as number,
            color: _randomColor(index),
        } as IGaugeChartSegment;
    });
}

function getDummyAction(key: string, index: number): IGaugeChartSegment {
    return {
        id: key,
        key: key,
        legend: key,
        color: _randomColor(index),
        size: 30,
    } as IGaugeChartSegment;
}
