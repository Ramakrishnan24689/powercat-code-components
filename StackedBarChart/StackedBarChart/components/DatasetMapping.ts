import { ItemColumns } from '../ManifestConstants';
import { DataVizPalette, IChartDataPoint, IChartProps, getColorFromToken } from '@fluentui/react-charting';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    const colorGroup = _colors[index % _colors.length];
    return getColorFromToken(colorGroup[Math.floor(Math.random() * colorGroup.length)]);
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    title: string,
    ariaLabel: string,
    customColor: boolean,
): IChartProps {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return getDummyData();
    }
    return {
        chartTitle: title,
        chartDataAccessibilityData: { ariaLabel: ariaLabel },
        chartData: dataset.sortedRecordIds.map((id, index) => {
            const record = dataset.records[id];
            const key = record.getValue(ItemColumns.Key) as string;
            const keyIndex: Record<string, number> = {};
            keyIndex[key] = (keyIndex[key] || 0) + 1;
            return createChartProps(
                record,
                key + (keyIndex[key] > 1 ? `_${keyIndex[key] - 1}` : ''),
                index,
                customColor,
            );
        }),
    } as IChartProps;
}

function createChartProps(
    record: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord,
    key: string,
    index: number,
    customColor: boolean,
): IChartDataPoint {
    return {
        legend: record.getValue(ItemColumns.Title) as string,
        data: record.getValue(ItemColumns.Value) as number,
        color: customColor ? getColorFromToken(record.getValue(ItemColumns.Color) as string) : _randomColor(index),
        xAxisCalloutData: record.getValue(ItemColumns.Callout) as string,
    } as IChartDataPoint;
}

function getDummyData(): IChartProps {
    return {
        chartTitle: 'StackChart',
        chartData: [
            {
                data: 100,
                xAxisCalloutData: '1',
                legend: 'Item ' + 1,
                color: _randomColor(1),
            },
            {
                data: 200,
                xAxisCalloutData: '2',
                legend: 'Item ' + 2,
                color: _randomColor(2),
            },
            {
                data: 300,
                xAxisCalloutData: '3',
                legend: 'Item ' + 3,
                color: _randomColor(3),
            },
        ],
    } as IChartProps;
}
