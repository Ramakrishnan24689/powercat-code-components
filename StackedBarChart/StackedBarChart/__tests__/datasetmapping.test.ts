import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';
import { getChartDataFromDataset } from '../components/DatasetMapping';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());
    const stackBarLabel = 'Stack Bar Chart';
    it('returns correct props when custom color is false', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Title]: 'Item 1',
                [ItemColumns.Value]: 30,
                [ItemColumns.Color]: 'blue',
            }),
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Title]: 'Item 1',
                [ItemColumns.Value]: 20,
                [ItemColumns.Color]: 'blue',
            }),
        ];
        const actions = getChartDataFromDataset(new MockDataSet(items), stackBarLabel,  stackBarLabel, false);
        expect(actions).toMatchSnapshot();
    });

    it('returns correct props when custom color is true ', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Title]: 'Item 1',
                [ItemColumns.Value]: 30,
                [ItemColumns.Color]: 'blue',
            }),
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Title]: 'Item 1',
                [ItemColumns.Value]: 20,
                [ItemColumns.Color]: 'blue',
            }),
        ];
        const actions = getChartDataFromDataset(new MockDataSet(items), stackBarLabel,  stackBarLabel, false);
        expect(actions).toMatchSnapshot();
    });
});
