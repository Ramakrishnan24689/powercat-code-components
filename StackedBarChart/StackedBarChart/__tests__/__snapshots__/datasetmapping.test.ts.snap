// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props when custom color is false 1`] = `
Object {
  "chartData": Array [
    Object {
      "color": "#2aa0a4",
      "data": 20,
      "legend": "Item 1",
      "xAxisCalloutData": undefined,
    },
    Object {
      "color": "#57811b",
      "data": 20,
      "legend": "Item 1",
      "xAxisCalloutData": undefined,
    },
  ],
  "chartDataAccessibilityData": Object {
    "ariaLabel": "Stack Bar Chart",
  },
  "chartTitle": "Stack Bar Chart",
}
`;

exports[`DatasetMapping returns correct props when custom color is true  1`] = `
Object {
  "chartData": Array [
    Object {
      "color": "#2aa0a4",
      "data": 20,
      "legend": "Item 1",
      "xAxisCalloutData": undefined,
    },
    Object {
      "color": "#13a10e",
      "data": 20,
      "legend": "Item 1",
      "xAxisCalloutData": undefined,
    },
  ],
  "chartDataAccessibilityData": Object {
    "ariaLabel": "Stack Bar Chart",
  },
  "chartTitle": "Stack Bar Chart",
}
`;
